### PR TITLE
EWPP-3390: Active revision submodule.

### PR DESCRIPTION
**Pull request by upchuk**

### CODE REVIEW

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/oe_translation_active_revision.links.action.yml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+oe_translation_active_revision.add_form:
+  title: 'Add activerevision'
```
Comment (sergepavle) [inline modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/oe_translation_active_revision.links.action.yml]:
```suggestion
  title: 'Add active revision'
```

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/Entity/ActiveRevision.php [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+use Drupal\Core\Field\BaseFieldDefinition;
+use Drupal\oe_translation_active_revision\ActiveRevisionInterface;
+use Drupal\user\EntityOwnerTrait;
+
+/**
+ * Defines the oe_translation_active_revision entity class.
+ *
+ * @ContentEntityType(
+ *   id = "oe_translation_active_revision",
+ *   label = @Translation("ActiveRevision"),
```
Comment (sergepavle) [inline modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/Entity/ActiveRevision.php]: Is it intentional? 
```suggestion
 *   label = @Translation("Active Revision"),
```

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/Entity/ActiveRevision.php [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+ *   label_collection = @Translation("ActiveRevisions"),
+ *   label_singular = @Translation("activerevision"),
+ *   label_plural = @Translation("activerevisions"),
+ *   label_count = @PluralTranslation(
+ *     singular = "@count activerevisions",
```
Comment (sergepavle) [inline modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/Entity/ActiveRevision.php]:
```suggestion
 *     singular = "@count active revision",
```

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/Entity/ActiveRevision.php [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+use Drupal\oe_translation_active_revision\ActiveRevisionInterface;
+use Drupal\user\EntityOwnerTrait;
+
+/**
+ * Defines the oe_translation_active_revision entity class.
+ *
+ * @ContentEntityType(
+ *   id = "oe_translation_active_revision",
+ *   label = @Translation("ActiveRevision"),
+ *   label_collection = @Translation("ActiveRevisions"),
```
Comment (sergepavle) [inline modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/Entity/ActiveRevision.php]:
```suggestion
 *   label_collection = @Translation("Active Revisions"),
```

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/Entity/ActiveRevision.php [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+ *   label_singular = @Translation("activerevision"),
```
Comment (sergepavle) [inline modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/Entity/ActiveRevision.php]:
```suggestion
 *   label_singular = @Translation("active revision"),
```

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/Entity/ActiveRevision.php [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+ *   label_plural = @Translation("activerevisions"),
```
Comment (sergepavle) [inline modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/Entity/ActiveRevision.php]:
```suggestion
 *   label_plural = @Translation("active revisions"),
```

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/Entity/ActiveRevision.php [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+ *   label_count = @PluralTranslation(
+ *     singular = "@count activerevisions",
+ *     plural = "@count activerevisions",
```
Comment (sergepavle) [inline modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/Entity/ActiveRevision.php]:
```suggestion
 *     plural = "@count active revisions",
```

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/Entity/ActiveRevision.php [hunk 8]
```diff
@@ -0,0 +1,145 @@
+<?php
```
Comment (sergepavle) [inline modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/Entity/ActiveRevision.php:1]:
```suggestion
<?php

declare(strict_types = 1);
```

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/Form/ActiveRevisionSettingsForm.php [hunk 9]
```diff
@@ -0,0 +1,48 @@
+<?php
+
```
Comment (sergepavle) [inline modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/Form/ActiveRevisionSettingsForm.php:2]:
```suggestion
declare(strict_types = 1);

```

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/Form/ActiveRevisionSettingsForm.php [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+namespace Drupal\oe_translation_active_revision\Form;
+
+use Drupal\Core\Form\FormBase;
+use Drupal\Core\Form\FormStateInterface;
+
+/**
+ * Configuration form for an activerevision entity type.
```
Comment (sergepavle) [inline modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/Form/ActiveRevisionSettingsForm.php]:
```suggestion
 * Configuration form for an active revision entity type.
```

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/Form/ActiveRevisionForm.php [hunk 11]
```diff
@@ -0,0 +1,44 @@
+<?php
+
```
Comment (sergepavle) [inline modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/Form/ActiveRevisionForm.php:2]:
```suggestion
declare(strict_types = 1);

```

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/Entity/ActiveRevision.php [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+ *     "id" = "id",
+ *     "label" = "id",
+ *     "uuid" = "uuid",
+ *     "owner" = "uid",
+ *   },
+ *   links = {
+ *     "collection" = "/admin/content/oe-translation-active-revision",
+ *     "add-form" = "/admin/structure/active-revisions/add",
+ *     "canonical" = "/admin/structure/active-revisions/{oe_translation_active_revision}",
+ *     "edit-form" = "/admin/structure/active-revisions/{oe_translation_active_revision}",
```
Comment (sergepavle) [inline modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/Entity/ActiveRevision.php:53]: Canonical and edit form classes are the same. Is it intentional?

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/Entity/ActiveRevision.php [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+ *       "html" = "Drupal\oe_translation_active_revision\Routing\ActiveRevisionHtmlRouteProvider",
+ *     }
+ *   },
+ *   base_table = "oe_translation_active_revision",
+ *   admin_permission = "administer active revision settings",
+ *   entity_keys = {
+ *     "id" = "id",
+ *     "label" = "id",
+ *     "uuid" = "uuid",
+ *     "owner" = "uid",
```
Comment (sergepavle) [inline modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/Entity/ActiveRevision.php:47]: Is it planned to take into consideration ownership of the entity?

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/ActiveRevisionStorage.php [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+
+namespace Drupal\oe_translation_active_revision;
+
+use Drupal\Core\Cache\CacheableMetadata;
+use Drupal\Core\Entity\ContentEntityInterface;
+use Drupal\Core\Entity\Sql\SqlContentEntityStorage;
+use Drupal\oe_translation_active_revision\Plugin\Field\FieldType\LanguageWithEntityRevisionItem;
+
+/**
+ * Defines the storage handler class for Active Revision entities..
```
Comment (sergepavle) [inline modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/ActiveRevisionStorage.php]:
```suggestion
 * Defines the storage handler class for Active Revision entities.
```

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/config/install/field.storage.oe_translation_active_revision.field_language_revision.yml [hunk 15]
```diff
@@ -0,0 +1,17 @@
+langcode: en
+status: true
+dependencies:
+  module:
+    - oe_translation_active_revision
+id: oe_translation_active_revision.field_language_revision
```
Comment (sergepavle) [inline modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/config/install/field.storage.oe_translation_active_revision.field_language_revision.yml:6]: Why this field cannot be defined in the entity as the base field?

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/tests/src/FunctionalJavascript/ActiveRevisionTest.php [hunk 16]
```diff
@@ -0,0 +1,573 @@
...
+      $this->assertFalse($this->getSession()->getPage()->findField('The new version does NOT need NEW translations (there have been only changes to non-translatable fields)')->isChecked());
+      $this->assertTrue($this->getSession()->getPage()->findField('Keep current translations until new ones are synchronised.')->isVisible());
+      $this->assertTrue($this->getSession()->getPage()->findField('Keep current translations until new ones are synchronised.')->isChecked());
+      $this->assertTrue($this->getSession()->getPage()->findField('Delete current translations until new ones are synchronized')->isVisible());
+      $this->assertFalse($this->getSession()->getPage()->findField('Delete current translations until new ones are synchronized')->isChecked());
+
+      // Assert the state changes the other options.
+      $this->getSession()->getPage()->findField('The new version does NOT need NEW translations (there have been only changes to non-translatable fields)')->click();
+      $this->assertFalse($this->getSession()->getPage()->findField('Keep current translations until new ones are synchronised.')->isVisible());
```
Comment (sergepavle) [inline modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/tests/src/FunctionalJavascript/ActiveRevisionTest.php:185]: We may again assert that `Delete current translations until new ones are synchronized` is also not visible.
```suggestion
      $this->assertFalse($this->getSession()->getPage()->findField('Keep current translations until new ones are synchronised.')->isVisible());
      $this->assertFalse($this->getSession()->getPage()->findField('Delete current translations until new ones are synchronized')->isVisible());
```

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/oe_translation_active_revision.routing.yml [hunk 17]
```diff
@@ -0,0 +1,7 @@
+entity.oe_translation_active_revision.settings:
+  path: 'admin/structure/oe-translation-active-revision'
+  defaults:
+    _form: '\Drupal\oe_translation_active_revision\Form\ActiveRevisionSettingsForm'
+    _title: 'ActiveRevision'
```
Comment (sergepavle) [inline modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/oe_translation_active_revision.routing.yml]:
```suggestion
    _title: 'Active Revision'
```

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/ParamConverter/ActiveRevisionNodeConverter.php [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+    $this->languageManager = $language_manager;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function applies($definition, $name, Route $route) {
+    // Handle the case for the Node param only on the node canonical route and
+    // the latest revision route.
+    if (!empty($definition['type']) && strpos($definition['type'], 'entity:') === 0) {
```
Comment (sergepavle) [inline modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/ParamConverter/ActiveRevisionNodeConverter.php]:
```suggestion
    if (!empty($definition['type']) && str_starts_with($definition['type'], 'entity:')) {
```

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/ParamConverter/ActiveRevisionConverterTrait.php [hunk 19]
```diff
@@ -0,0 +1,92 @@
+<?php
+
+declare(strict_types=1);
```
Comment (sergepavle) [inline modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/ParamConverter/ActiveRevisionConverterTrait.php]:
```suggestion
declare(strict_types = 1);
```

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/ParamConverter/ActiveRevisionConverterTrait.php [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+
+/**
+ * Helper for performing the param conversion based on the active revision.
+ */
+trait ActiveRevisionConverterTrait {
+
+  /**
+   * Performs the conversion.
+   */
+  public function doConvert($value, $definition, $name, array $defaults) {
```
Comment (sergepavle) [inline modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/ParamConverter/ActiveRevisionConverterTrait.php:31]: Missed return type and PhpDoc comment.

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/ParamConverter/ActiveRevisionConverterTrait.php [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+    $contexts_repository = \Drupal::service('context.repository');
```
Comment (sergepavle) [inline modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/ParamConverter/ActiveRevisionConverterTrait.php]: Is it possible to do injection of `context.repository` service?

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/ParamConverter/ActiveRevisionConverterTrait.php [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+  /**
+   * Retrieves the current content language from the specified contexts.
+   *
+   * @param \Drupal\Core\Plugin\Context\ContextInterface[] $contexts
+   *   An array of context items.
+   *
+   * @return string|null
+   *   A language code or NULL if no language context was provided.
+   */
+  protected function getContentLanguageFromContexts(array $contexts) {
```
Comment (sergepavle) [inline modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/ParamConverter/ActiveRevisionConverterTrait.php]:
```suggestion
  protected function getContentLanguageFromContexts(array $contexts): string {
```

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/ParamConverter/ActiveRevisionConverterTrait.php [hunk 23]
```diff
@@ -0,0 +1,92 @@
...
+  }
+
+  /**
+   * Retrieves the current content language from the specified contexts.
+   *
+   * @param \Drupal\Core\Plugin\Context\ContextInterface[] $contexts
+   *   An array of context items.
+   *
+   * @return string|null
+   *   A language code or NULL if no language context was provided.
```
Comment (sergepavle) [inline modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/ParamConverter/ActiveRevisionConverterTrait.php]: But we anyway if don't have anything in contexts returning default language. Is it intentional?

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/ParamConverter/ActiveRevisionNodeRevisionConverter.php [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+
+/**
+ * Param converter for the active revision for the node revision.
+ *
+ * We need this because the entity view controller, after building the entity
+ * render array and all the view and build hooks have passed, it just
+ * replaces back the rendered entity with the initial one. Because normally,
+ * for other view modes, we use
+ * oe_translation_active_revision_entity_build_defaults_alter() to switch out
+ * the entity revision.
```
Comment (sergepavle) [inline modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/ParamConverter/ActiveRevisionNodeRevisionConverter.php:22]: It is not very clear from comment difference with `modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/ParamConverter/ActiveRevisionNodeConverter.php`

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/Routing/ActiveRevisionHtmlRouteProvider.php [hunk 25]
```diff
@@ -0,0 +1,20 @@
+<?php
+
```
Comment (sergepavle) [inline modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/Routing/ActiveRevisionHtmlRouteProvider.php:2]:
```suggestion
declare(strict_types = 1);

```

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/tests/src/FunctionalJavascript/ActiveRevisionTest.php [hunk 26]
```diff
@@ -0,0 +1,573 @@
...
+      $node = $this->moderateNode($node, 'published');
+
+      $node->addTranslation('fr', ['title' => 'My FR node']);
+      $node->save();
+
+      // Keep track of the version 1 revision ID.
+      $version_one_revision_id = $node->getRevisionId();
+
+      $this->drupalGet($node->toUrl());
+      $this->assertSession()->pageTextContains('My node');
```
Comment (sergepavle) [inline modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/tests/src/FunctionalJavascript/ActiveRevisionTest.php:153]: Here we may again assert that we still don't see form elements for mapping the active revision.

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/ActiveRevisionInterface.php [hunk 27]
```diff
@@ -0,0 +1,14 @@
+<?php
+
```
Comment (sergepavle) [inline modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/ActiveRevisionInterface.php:2]:
```suggestion

declare(strict_types = 1);

```

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/ActiveRevisionListBuilder.php [hunk 28]
```diff
@@ -0,0 +1,94 @@
+<?php
+
```
Comment (sergepavle) [inline modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/ActiveRevisionListBuilder.php:2]:
```suggestion

declare(strict_types = 1);

```

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/ActiveRevisionListBuilder.php [hunk 29]
```diff
@@ -0,0 +1,94 @@
...
+use Drupal\Core\Datetime\DateFormatterInterface;
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\Core\Entity\EntityListBuilder;
+use Drupal\Core\Entity\EntityStorageInterface;
+use Drupal\Core\Entity\EntityTypeInterface;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Provides a list controller for the activerevision entity type.
```
Comment (sergepavle) [inline modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/ActiveRevisionListBuilder.php]:
```suggestion
 * Provides a list controller for the active revision entity type.
```

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/ActiveRevisionInterface.php [hunk 30]
```diff
@@ -0,0 +1,14 @@
...
+namespace Drupal\oe_translation_active_revision;
+
+use Drupal\Core\Entity\ContentEntityInterface;
+use Drupal\Core\Entity\EntityChangedInterface;
+use Drupal\user\EntityOwnerInterface;
+
+/**
+ * Provides an interface defining an activerevision entity type.
```
Comment (sergepavle) [inline modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/ActiveRevisionInterface.php]:
```suggestion
 * Provides an interface defining an active revision entity type.
```

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/oe_translation_active_revision.module [hunk 31]
```diff
@@ -0,0 +1,286 @@
...
+  ];
+  $form['translation_options_wrapper']['translation_option_needs_new_wrapper']['translation_option_needs_new'] = [
+    '#type' => 'radios',
+    '#options' => [
+      'use_previous_version_translations' => t('Keep current translations until new ones are synchronised.'),
+      'delete_translations' => t('Delete current translations until new ones are synchronized'),
+    ],
+    '#default_value' => 'use_previous_version_translations',
+  ];
+  $form['translation_options_wrapper']['translation_option_needs_no_ew_wrapper'] = [
```
Comment (sergepavle) [inline modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/oe_translation_active_revision.module]:
```suggestion
  $form['translation_options_wrapper']['translation_option_needs_no_new_wrapper'] = [
```

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/Entity/ActiveRevision.php (hunk 12 above)
Comment (upchuk) [reply modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/Entity/ActiveRevision.php:53]: Yes, we don't need any other type of canonical. Like media.

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/Entity/ActiveRevision.php (hunk 13 above)
Comment (upchuk) [reply modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/Entity/ActiveRevision.php:47]: No, I just left there the defaults.

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/config/install/field.storage.oe_translation_active_revision.field_language_revision.yml (hunk 15 above)
Comment (upchuk) [reply modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/config/install/field.storage.oe_translation_active_revision.field_language_revision.yml:6]: It's just easier like this. It's not uncommon. Doesn't matter much, but safer with config fields, support is not always 100% 1 to 1.

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/ParamConverter/ActiveRevisionConverterTrait.php (hunk 20 above)
Comment (upchuk) [reply modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/ParamConverter/ActiveRevisionConverterTrait.php:31]: I put a phpdoc but left out the return types to match the parent caller which is from core.

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/ParamConverter/ActiveRevisionConverterTrait.php (hunk 23 above)
Comment (upchuk) [reply modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/ParamConverter/ActiveRevisionConverterTrait.php]: Comment was wrong, fixed it.

File: modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/ParamConverter/ActiveRevisionNodeRevisionConverter.php (hunk 24 above)
Comment (upchuk) [reply modules/oe_translation_corporate_workflow/modules/oe_translation_active_revision/src/ParamConverter/ActiveRevisionNodeRevisionConverter.php:22]: The comment doesn't describe what the difference is between `ActiveRevisionNodeConverter` and `ActiveRevisionNodeRevisionConverter`, but why we need them. Normally, we handle things in oe_translation_active_revision_entity_build_defaults_alter but on the entity view controller that doesnt work. As for the reason we need both, you can see in the applies which paths/routes they are applicable to.

**Review by sergepavle — APPROVED** [review]

**Review by hernani — APPROVED** [review]

**Pull request closed without merging**